src/pybind/mgr/cephadm/module.py

```diff
@@ -1598,14 +1598,21 @@ def _daemon_action(self, daemon_type: str, daemon_id: str, host: str, action: st
 
         self._daemon_action_set_image(action, image, daemon_type, daemon_id)
 
-        if action == 'redeploy':
-            if self.daemon_is_self(daemon_type, daemon_id):
-                self.mgr_service.fail_over()
-                return ''  # unreachable
-            # stop, recreate the container+unit, then restart
-            return CephadmServe(self)._create_daemon(daemon_spec)
-        elif action == 'reconfig':
-            return CephadmServe(self)._create_daemon(daemon_spec, reconfig=True)
+        if action == 'redeploy' and self.daemon_is_self(daemon_type, daemon_id):
+            self.mgr_service.fail_over()
+            return ''  # unreachable
+
+        if action == 'redeploy' or action == 'reconfig':
+            if daemon_type != 'osd':
+                try:
+                    daemon_spec = self.cephadm_services[daemon_type_to_service(
+                        daemon_type)].prepare_create(daemon_spec)
+                except (RuntimeError, OrchestratorError) as e:
+                    msg = "Failed to {} daemon {} on {}: {}".format(
+                        action, daemon_spec.name(), daemon_spec.host, e)
+                    self.log.info(msg)
+                    return msg
```

Comment:
Unfortunately, we have the irritating situation that we have two ways to deal with errors in cephadm:

1. return errors as strings
2. return errors as `OrchestratorError` exceptions. 


After having a close look, it turned out that caching exceptions here won't work as intended:

Let's have alook at serve.py and upgrade.py:

https://github.com/ceph/ceph/blob/29f167eecf72094c33580fd403c9c18e82f3e93a/src/pybind/mgr/cephadm/serve.py#L664-L676

https://github.com/ceph/ceph/blob/29f167eecf72094c33580fd403c9c18e82f3e93a/src/pybind/mgr/cephadm/upgrade.py#L495-L512

Have a look at those except blocks: Thy both expect `daemon_action` to raise exceptions in case of errors. 

after your PR, the upgrade.py would think that the call to `_daemon_action` would have succeeded, even though `_create_daemon` was never called! What do you think of simply removing the `try-except` block?

Reply:
Removing the try-except block works for me. I'll make that change.